# 01 Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02 Importing citibike dataframe

In [3]:
# Creating path first
path = r'/Users/jaksdt/Desktop/Data analytics stuff/Data immersion/Exercises 1 - 10/Section 6'

In [4]:
# importing data frame
df = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'citibike.csv'), index_col = False)

In [5]:
df

trip_id  bike_id weekday  start_hour           start_time  \
0      LnQzQk    16013     Mon          18  2013-09-09 18:18:55   
1      IL9boN    15230     Thu          18  2013-09-12 18:38:53   
2      46clGB    17942     Wed          19  2013-09-18 19:44:04   
3      v7vdFt    19683     Sat          11  2013-09-28 11:54:37   
4      VGBsb5    18024     Sat          18  2013-09-07 18:08:22   
...       ...      ...     ...         ...                  ...   
49995  Fb2bmC    17988     Sat          18  2013-09-21 18:39:20   
49996  XGB7JU    18458     Sat          17  2013-09-14 17:32:26   
49997  vFs2HS    15054     Sat          18  2013-09-07 18:08:21   
49998  xxacSt    16763     Sat          13  2013-09-07 13:25:35   
49999  4ZmicF    16794     Mon          21  2013-09-23 21:54:36   

       start_station_id       start_station_name  start_station_latitude  \
0                   523          W 38 St & 8 Ave               40.754666   
1                   257  Lispenard St & Broadway               40.719392   
2                   479          9 Ave & W 45 St               40.760193   
3                   527          E 33 St & 1 Ave               40.743156   
4                   521          8 Ave & W 31 St               40.750450   
...                 ...                      ...                     ...   
49995               479          9 Ave & W 45 St               40.760193   
49996               261     Johnson St & Gold St               40.694749   
49997               402       Broadway & E 22 St               40.740343   
49998               271   Ashland Pl & Hanson Pl               40.685282   
49999               515         W 43 St & 10 Ave               40.760094   

       start_station_longitude             end_time  end_station_id  \
0                   -73.991382  2013-09-09 18:35:28             334   
1                   -74.002472  2013-09-12 18:48:34             236   
2                   -73.991255  2013-09-18 19:50:05             513   
3                   -73.974347  2013-09-28 12:03:58             441   
4                   -73.994811  2013-09-07 18:46:38             476   
...                        ...                  ...             ...   
49995               -73.991255  2013-09-21 20:08:59             478   
49996               -73.983625  2013-09-14 17:52:17             151   
49997               -73.989551  2013-09-07 18:25:51             524   
49998               -73.978058  2013-09-07 14:14:04             373   
49999               -73.994618  2013-09-23 22:00:46             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year  \
0                 -73.997262            993      Subscriber      1968.0   
1                 -73.987140            581      Subscriber      1983.0   
2                 -73.988639            361      Subscriber      1989.0   
3                 -73.967416            561      Subscriber      1988.0   
4                 -73.979661           2296  Non-Subscriber         NaN   
...                      ...            ...             ...         ...   
49995             -73.998842           2697  Non-Subscriber         NaN   
49996             -73.997203           1191      Subscriber      1970.0   
49997             -73.98316

In [6]:
df.shape

(50000, 18)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   trip_id                  50000 non-null  object 
 1   bike_id                  50000 non-null  int64  
 2   weekday                  50000 non-null  object 
 3   start_hour               50000 non-null  int64  
 4   start_time               50000 non-null  object 
 5   start_station_id         50000 non-null  int64  
 6   start_station_name       50000 non-null  object 
 7   start_station_latitude   50000 non-null  float64
 8   start_station_longitude  50000 non-null  float64
 9   end_time                 50000 non-null  object 
 10  end_station_id           50000 non-null  int64  
 11  end_station_name         50000 non-null  object 
 12  end_station_latitude     50000 non-null  float64
 13  end_station_longitude    50000 non-null  float64
 14  trip_duration         

In [9]:
df.describe()

bike_id    start_hour  start_station_id  start_station_latitude  \
count  50000.000000  50000.000000      50000.000000            50000.000000   
mean   17615.269360     14.145240        443.321500               40.734170   
std     1675.407446      4.860541        356.559925                0.019911   
min    14556.000000      0.000000         72.000000               40.680342   
25%    16188.000000     10.000000        304.000000               40.720196   
50%    17584.000000     15.000000        402.000000               40.735877   
75%    19014.000000     18.000000        484.000000               40.750020   
max    20642.000000     23.000000       3002.000000               40.770513   

       start_station_longitude  end_station_id  end_station_latitude  \
count             50000.000000    50000.000000          50000.000000   
mean                -73.991109      442.539700             40.733859   
std                   0.012555      355.756022              0.019885   
min                 -74.017134       72.000000             40.680342   
25%                 -74.000271      304.000000             40.720196   
50%                 -73.990765      402.000000             40.735354   
75%                 -73.981923      483.000000             40.749013   
max                 -73.950048     3002.000000             40.770513   

       end_station_longitude  trip_duration    birth_year        gender  
count           50000.000000   50000.000000  43021.000000  50000.000000  
mean              -73.991351     838.982900   1975.627786      1.073540  
std                 0.012569     573.663997     11.089001      0.589389  
min               -74.017134      60.000000   1899.000000      0.000000  
25%               -74.001547     417.000000   1968.000000      1.000000  
50%               -73.991218     672.000000   1978.000000      1.000000  
75%               -73.982050    1112.000000   1984.000000      1.000000  
max               -73.950048    2697.000000   1997.000000      2.000000

# 03. Data wrangling

In [10]:
# Checking for missing values
df.isnull().sum()

trip_id                       0
bike_id                       0
weekday                       0
start_hour                    0
start_time                    0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_time                      0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
trip_duration                 0
subscriber                    0
birth_year                 6979
gender                        0
dtype: int64

### Only birth year has missing values. We can't drop them, because there are too many. Those rows will contain valuable information needed for our analysis.

In [11]:
# Dropping unnecessary column
df = df.drop(columns = ['trip_id'])

In [12]:
df

bike_id weekday  start_hour           start_time  start_station_id  \
0        16013     Mon          18  2013-09-09 18:18:55               523   
1        15230     Thu          18  2013-09-12 18:38:53               257   
2        17942     Wed          19  2013-09-18 19:44:04               479   
3        19683     Sat          11  2013-09-28 11:54:37               527   
4        18024     Sat          18  2013-09-07 18:08:22               521   
...        ...     ...         ...                  ...               ...   
49995    17988     Sat          18  2013-09-21 18:39:20               479   
49996    18458     Sat          17  2013-09-14 17:32:26               261   
49997    15054     Sat          18  2013-09-07 18:08:21               402   
49998    16763     Sat          13  2013-09-07 13:25:35               271   
49999    16794     Mon          21  2013-09-23 21:54:36               515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude             end_time  end_station_id  \
0                   -73.991382  2013-09-09 18:35:28             334   
1                   -74.002472  2013-09-12 18:48:34             236   
2                   -73.991255  2013-09-18 19:50:05             513   
3                   -73.974347  2013-09-28 12:03:58             441   
4                   -73.994811  2013-09-07 18:46:38             476   
...                        ...                  ...             ...   
49995               -73.991255  2013-09-21 20:08:59             478   
49996               -73.983625  2013-09-14 17:52:17             151   
49997               -73.989551  2013-09-07 18:25:51             524   
49998               -73.978058  2013-09-07 14:14:04             373   
49999               -73.994618  2013-09-23 22:00:46             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year  \
0                 -73.997262            993      Subscriber      1968.0   
1                 -73.987140            581      Subscriber      1983.0   
2                 -73.988639            361      Subscriber      1989.0   
3                 -73.967416            561      Subscriber      1988.0   
4                 -73.979661           2296  Non-Subscriber         NaN   
...                      ...            ...             ...         ...   
49995             -73.998842           2697  Non-Subscriber         NaN   
49996             -73.997203           1191      Subscriber      1970.0   
49997             -73.983169           1050      Subscriber      1961.0   
49998             -73.953820           2697  Non-

In [13]:
# Converting gender from int to string
df['gender'] = df['gender'].astype('str')

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   bike_id                  50000 non-null  int64  
 1   weekday                  50000 non-null  object 
 2   start_hour               50000 non-null  int64  
 3   start_time               50000 non-null  object 
 4   start_station_id         50000 non-null  int64  
 5   start_station_name       50000 non-null  object 
 6   start_station_latitude   50000 non-null  float64
 7   start_station_longitude  50000 non-null  float64
 8   end_time                 50000 non-null  object 
 9   end_station_id           50000 non-null  int64  
 10  end_station_name         50000 non-null  object 
 11  end_station_latitude     50000 non-null  float64
 12  end_station_longitude    50000 non-null  float64
 13  trip_duration            50000 non-null  int64  
 14  subscriber            

In [15]:
# Checking and creating subset for anyone older than 95
df_sub = df.loc[df['birth_year'] < 1917]

In [16]:
df_sub

bike_id weekday  start_hour           start_time  start_station_id  \
563      19507     Mon          12  2013-09-23 12:43:58               305   
1371     16277     Thu          11  2013-09-05 11:20:55               500   
1431     14638     Thu           7  2013-09-26 07:35:18               427   
1754     17594     Wed          17  2013-09-25 17:00:04               476   
4823     19804     Thu           7  2013-09-12 07:36:39               515   
7502     17490     Mon           7  2013-09-30 07:38:46               498   
9644     15732     Sun          11  2013-09-08 11:11:35               515   
12926    14944     Fri          20  2013-09-20 20:16:18               427   
13081    20512     Tue           7  2013-09-10 07:53:12               471   
13575    18462     Wed           8  2013-09-11 08:40:59               514   
16135    15685     Tue          18  2013-09-10 18:53:53               268   
16252    14581     Thu          17  2013-09-26 17:48:15               358   
17130    19744     Thu           7  2013-09-12 07:12:03               427   
18332    19746     Wed          11  2013-09-11 11:21:53               327   
19824    15104     Thu          18  2013-09-05 18:11:52               540   
28568    18769     Sun          11  2013-09-22 11:07:36              2004   
37638    19667     Thu          19  2013-09-05 19:42:45               306   
39863    19775     Sun          13  2013-09-08 13:37:24               327   
40359    19749     Thu           7  2013-09-19 07:58:17               147   
44309    15521     Wed          16  2013-09-11 16:22:35               225   
47919    19507     Sun          11  2013-09-22 11:20:44               347   
48989    16098     Tue          17  2013-09-03 17:44:39               485   
49684    15708     Mon          14  2013-09-30 14:54:49               328   

                  start_station_name  start_station_latitude  \
563                  E 58 St & 3 Ave               40.760958   
1371              Broadway & W 51 St               40.762288   
1431                        State St               40.702515   
1754                 E 31 St & 3 Ave               40.743943   
4823                W 43 St & 10 Ave               40.760094   
7502              Broadway & W 32 St               40.748549   
9644                W 43 St & 10 Ave               40.760094   
12926                       State St               40.702515   
13081        Grand St & Havemeyer St               40.712868   
13575               12 Ave & W 40 St               40.760875   
16135          Howard St & Centre St               40.719105   
16252  Christopher St & Greenwich St               40.732916   
17130                       State St               40.702515   
18332       Vesey Pl & River Terrace               40.715338   
19824        Lexington Ave & E 26 St               40.741473   
28568              6 Ave & Broome St               40.724399   
37638           Cliff St & Fulton St               40.708235   
39863       Vesey Pl & River Terrace               40.715338   
40359       Greenwich St & Warren St               40.715422   
44309        W 14 St & The High Line               40.741951   
47919       W Houston St & Hudson St               40.728739   
48989                W 37 St & 5 Ave               40.750380   
49684        Watts St & Greenwich St               40.724055   

       start_station_longitude             end_time  end_station_id  \
563                 -73.967245  2013-09-23 12:52:50            2006   
1371                -73.983362  2013-09-05 11:36:10             497   
1431                -74.014270  2013-09-26 08:00:02             212   
1754                -73.979661  2013-09-25 17:09:33             435   
4823                -73.994618  2013-09-12 07:55:02             268   
7502                -73.988084  2013-09-30 08:01:17             340   
9644                -73.994618  2013-09-08 11:32:02             268   
12926               -74.014270  2013-09-20 20:58:45  

### Since Citi Bike's age requirement is 16 years to open up an account and 95 years old is pushing the limit, anyone older than that can be viewed as an outlier. So years 1899, 1900 etc cannot be changed to 1999 or 2000 etc thinking that it was a mistake, because you have to be 16 to open an account as this information was for September 2013. So these 23 entries will be dropped

In [17]:
# Dropping above rows
df = df.drop(df[df['birth_year'] < 1917].index)

In [18]:
df

bike_id weekday  start_hour           start_time  start_station_id  \
0        16013     Mon          18  2013-09-09 18:18:55               523   
1        15230     Thu          18  2013-09-12 18:38:53               257   
2        17942     Wed          19  2013-09-18 19:44:04               479   
3        19683     Sat          11  2013-09-28 11:54:37               527   
4        18024     Sat          18  2013-09-07 18:08:22               521   
...        ...     ...         ...                  ...               ...   
49995    17988     Sat          18  2013-09-21 18:39:20               479   
49996    18458     Sat          17  2013-09-14 17:32:26               261   
49997    15054     Sat          18  2013-09-07 18:08:21               402   
49998    16763     Sat          13  2013-09-07 13:25:35               271   
49999    16794     Mon          21  2013-09-23 21:54:36               515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude             end_time  end_station_id  \
0                   -73.991382  2013-09-09 18:35:28             334   
1                   -74.002472  2013-09-12 18:48:34             236   
2                   -73.991255  2013-09-18 19:50:05             513   
3                   -73.974347  2013-09-28 12:03:58             441   
4                   -73.994811  2013-09-07 18:46:38             476   
...                        ...                  ...             ...   
49995               -73.991255  2013-09-21 20:08:59             478   
49996               -73.983625  2013-09-14 17:52:17             151   
49997               -73.989551  2013-09-07 18:25:51             524   
49998               -73.978058  2013-09-07 14:14:04             373   
49999               -73.994618  2013-09-23 22:00:46             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year gender  
0                 -73.997262            993      Subscriber      1968.0      2  
1                 -73.987140            581      Subscriber      1983.0      1  
2                 -73.988639            361      Subscriber      1989.0      1  
3                 -73.967416            561      Subscriber      1988.0      2  
4                 -73.979661           2296  Non-Subscriber         NaN      0  
...                      ...            ...             ...         ...    ...  
49995             -73.998842           2697  Non-Subscriber         NaN      0  
49996             -73.997203           1191      Subscriber      1970.0      2  
49997             -73.983169           1050      Subscriber      1961.

In [19]:
df.describe()

bike_id    start_hour  start_station_id  start_station_latitude  \
count  49977.000000  49977.000000      49977.000000            49977.000000   
mean   17615.280109     14.146027        443.310683               40.734171   
std     1675.220866      4.860599        356.565315                0.019911   
min    14556.000000      0.000000         72.000000               40.680342   
25%    16188.000000     10.000000        304.000000               40.720196   
50%    17584.000000     15.000000        402.000000               40.735877   
75%    19014.000000     18.000000        484.000000               40.750020   
max    20642.000000     23.000000       3002.000000               40.770513   

       start_station_longitude  end_station_id  end_station_latitude  \
count             49977.000000    49977.000000          49977.000000   
mean                -73.991106      442.519639             40.733861   
std                   0.012552      355.687559              0.019885   
min                 -74.017134       72.000000             40.680342   
25%                 -74.000271      304.000000             40.720196   
50%                 -73.990765      402.000000             40.735354   
75%                 -73.981923      483.000000             40.749013   
max                 -73.950048     3002.000000             40.770513   

       end_station_longitude  trip_duration    birth_year  
count           49977.000000   49977.000000  42998.000000  
mean              -73.991348     838.877624   1975.668078  
std                 0.012569     573.675762     10.954112  
min               -74.017134      60.000000   1917.000000  
25%               -74.001547     417.000000   1968.000000  
50%               -73.991218     672.000000   1978.000000  
75%               -73.982050    1112.000000   1984.000000  
max               -73.950048    2697.000000   1997.000000

In [20]:
# Replacing all NaN values in birth_year column with mean of 1976.0
df['birth_year'].fillna(1976.0, inplace = True)

In [21]:
# Checking for missing values
df.isnull().sum()

bike_id                    0
weekday                    0
start_hour                 0
start_time                 0
start_station_id           0
start_station_name         0
start_station_latitude     0
start_station_longitude    0
end_time                   0
end_station_id             0
end_station_name           0
end_station_latitude       0
end_station_longitude      0
trip_duration              0
subscriber                 0
birth_year                 0
gender                     0
dtype: int64

In [22]:
df

bike_id weekday  start_hour           start_time  start_station_id  \
0        16013     Mon          18  2013-09-09 18:18:55               523   
1        15230     Thu          18  2013-09-12 18:38:53               257   
2        17942     Wed          19  2013-09-18 19:44:04               479   
3        19683     Sat          11  2013-09-28 11:54:37               527   
4        18024     Sat          18  2013-09-07 18:08:22               521   
...        ...     ...         ...                  ...               ...   
49995    17988     Sat          18  2013-09-21 18:39:20               479   
49996    18458     Sat          17  2013-09-14 17:32:26               261   
49997    15054     Sat          18  2013-09-07 18:08:21               402   
49998    16763     Sat          13  2013-09-07 13:25:35               271   
49999    16794     Mon          21  2013-09-23 21:54:36               515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude             end_time  end_station_id  \
0                   -73.991382  2013-09-09 18:35:28             334   
1                   -74.002472  2013-09-12 18:48:34             236   
2                   -73.991255  2013-09-18 19:50:05             513   
3                   -73.974347  2013-09-28 12:03:58             441   
4                   -73.994811  2013-09-07 18:46:38             476   
...                        ...                  ...             ...   
49995               -73.991255  2013-09-21 20:08:59             478   
49996               -73.983625  2013-09-14 17:52:17             151   
49997               -73.989551  2013-09-07 18:25:51             524   
49998               -73.978058  2013-09-07 14:14:04             373   
49999               -73.994618  2013-09-23 22:00:46             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year gender  
0                 -73.997262            993      Subscriber      1968.0      2  
1                 -73.987140            581      Subscriber      1983.0      1  
2                 -73.988639            361      Subscriber      1989.0      1  
3                 -73.967416            561      Subscriber      1988.0      2  
4                 -73.979661           2296  Non-Subscriber      1976.0      0  
...                      ...            ...             ...         ...    ...  
49995             -73.998842           2697  Non-Subscriber      1976.0      0  
49996             -73.997203           1191      Subscriber      1970.0      2  
49997             -73.983169           1050      Subscriber      1961.

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 49977 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   bike_id                  49977 non-null  int64  
 1   weekday                  49977 non-null  object 
 2   start_hour               49977 non-null  int64  
 3   start_time               49977 non-null  object 
 4   start_station_id         49977 non-null  int64  
 5   start_station_name       49977 non-null  object 
 6   start_station_latitude   49977 non-null  float64
 7   start_station_longitude  49977 non-null  float64
 8   end_time                 49977 non-null  object 
 9   end_station_id           49977 non-null  int64  
 10  end_station_name         49977 non-null  object 
 11  end_station_latitude     49977 non-null  float64
 12  end_station_longitude    49977 non-null  float64
 13  trip_duration            49977 non-null  int64  
 14  subscriber               49

In [24]:
# Checking the days of the week
df['weekday'].unique()

array(['Mon', 'Thu', 'Wed', 'Sat', 'Fri', 'Sun', 'Tue'], dtype=object)

In [25]:
# Checking gender values
df['gender'].value_counts(dropna = False)

gender
1    32338
2    10658
0     6981
Name: count, dtype: int64

In [26]:
# Creating a function to re-categorize gender
def gender_category(row):
    if row['gender'] == '1':
        return 'Male'
    elif row['gender'] == '2':
        return 'Female'
    else: return 'Unknown'

In [27]:
# Applying the function
df['gender_cat'] = df.apply(gender_category, axis = 1)

In [28]:
df

bike_id weekday  start_hour           start_time  start_station_id  \
0        16013     Mon          18  2013-09-09 18:18:55               523   
1        15230     Thu          18  2013-09-12 18:38:53               257   
2        17942     Wed          19  2013-09-18 19:44:04               479   
3        19683     Sat          11  2013-09-28 11:54:37               527   
4        18024     Sat          18  2013-09-07 18:08:22               521   
...        ...     ...         ...                  ...               ...   
49995    17988     Sat          18  2013-09-21 18:39:20               479   
49996    18458     Sat          17  2013-09-14 17:32:26               261   
49997    15054     Sat          18  2013-09-07 18:08:21               402   
49998    16763     Sat          13  2013-09-07 13:25:35               271   
49999    16794     Mon          21  2013-09-23 21:54:36               515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude             end_time  end_station_id  \
0                   -73.991382  2013-09-09 18:35:28             334   
1                   -74.002472  2013-09-12 18:48:34             236   
2                   -73.991255  2013-09-18 19:50:05             513   
3                   -73.974347  2013-09-28 12:03:58             441   
4                   -73.994811  2013-09-07 18:46:38             476   
...                        ...                  ...             ...   
49995               -73.991255  2013-09-21 20:08:59             478   
49996               -73.983625  2013-09-14 17:52:17             151   
49997               -73.989551  2013-09-07 18:25:51             524   
49998               -73.978058  2013-09-07 14:14:04             373   
49999               -73.994618  2013-09-23 22:00:46             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year  \
0                 -73.997262            993      Subscriber      1968.0   
1                 -73.987140            581      Subscriber      1983.0   
2                 -73.988639            361      Subscriber      1989.0   
3                 -73.967416            561      Subscriber      1988.0   
4                 -73.979661           2296  Non-Subscriber      1976.0   
...                      ...            ...             ...         ...   
49995             -73.998842           2697  Non-Subscriber      1976.0   
49996             -73.997203           1191      Subscriber      1970.0   
49997             -73.983169           1050      Subscriber      1961.0   
49998             -73.953820           2697  Non-

In [29]:
# Changing start_time column from string to datetime
df['start_time'] = df['start_time'].astype('datetime64[ns]')

In [30]:
# Changing end_time column from string to datetime
df['end_time'] = df['end_time'].astype('datetime64[ns]')

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 49977 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   bike_id                  49977 non-null  int64         
 1   weekday                  49977 non-null  object        
 2   start_hour               49977 non-null  int64         
 3   start_time               49977 non-null  datetime64[ns]
 4   start_station_id         49977 non-null  int64         
 5   start_station_name       49977 non-null  object        
 6   start_station_latitude   49977 non-null  float64       
 7   start_station_longitude  49977 non-null  float64       
 8   end_time                 49977 non-null  datetime64[ns]
 9   end_station_id           49977 non-null  int64         
 10  end_station_name         49977 non-null  object        
 11  end_station_latitude     49977 non-null  float64       
 12  end_station_longitude    49977 non-nu

In [33]:
# Dropping gender column
df = df.drop(columns = ['gender'])

In [34]:
df

bike_id weekday  start_hour          start_time  start_station_id  \
0        16013     Mon          18 2013-09-09 18:18:55               523   
1        15230     Thu          18 2013-09-12 18:38:53               257   
2        17942     Wed          19 2013-09-18 19:44:04               479   
3        19683     Sat          11 2013-09-28 11:54:37               527   
4        18024     Sat          18 2013-09-07 18:08:22               521   
...        ...     ...         ...                 ...               ...   
49995    17988     Sat          18 2013-09-21 18:39:20               479   
49996    18458     Sat          17 2013-09-14 17:32:26               261   
49997    15054     Sat          18 2013-09-07 18:08:21               402   
49998    16763     Sat          13 2013-09-07 13:25:35               271   
49999    16794     Mon          21 2013-09-23 21:54:36               515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude            end_time  end_station_id  \
0                   -73.991382 2013-09-09 18:35:28             334   
1                   -74.002472 2013-09-12 18:48:34             236   
2                   -73.991255 2013-09-18 19:50:05             513   
3                   -73.974347 2013-09-28 12:03:58             441   
4                   -73.994811 2013-09-07 18:46:38             476   
...                        ...                 ...             ...   
49995               -73.991255 2013-09-21 20:08:59             478   
49996               -73.983625 2013-09-14 17:52:17             151   
49997               -73.989551 2013-09-07 18:25:51             524   
49998               -73.978058 2013-09-07 14:14:04             373   
49999               -73.994618 2013-09-23 22:00:46             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year  \
0                 -73.997262            993      Subscriber      1968.0   
1                 -73.987140            581      Subscriber      1983.0   
2                 -73.988639            361      Subscriber      1989.0   
3                 -73.967416            561      Subscriber      1988.0   
4                 -73.979661           2296  Non-Subscriber      1976.0   
...                      ...            ...             ...         ...   
49995             -73.998842           2697  Non-Subscriber      1976.0   
49996             -73.997203           1191      Subscriber      1970.0   
49997             -73.983169           1050      Subscriber      1961.0   
49998             -73.953820           2697  Non-Subscriber      1976.0  

In [35]:
# Renaming gender_cat column
df.rename(columns = {'gender_cat' : 'gender'}, inplace = True)

In [36]:
df

bike_id weekday  start_hour          start_time  start_station_id  \
0        16013     Mon          18 2013-09-09 18:18:55               523   
1        15230     Thu          18 2013-09-12 18:38:53               257   
2        17942     Wed          19 2013-09-18 19:44:04               479   
3        19683     Sat          11 2013-09-28 11:54:37               527   
4        18024     Sat          18 2013-09-07 18:08:22               521   
...        ...     ...         ...                 ...               ...   
49995    17988     Sat          18 2013-09-21 18:39:20               479   
49996    18458     Sat          17 2013-09-14 17:32:26               261   
49997    15054     Sat          18 2013-09-07 18:08:21               402   
49998    16763     Sat          13 2013-09-07 13:25:35               271   
49999    16794     Mon          21 2013-09-23 21:54:36               515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude            end_time  end_station_id  \
0                   -73.991382 2013-09-09 18:35:28             334   
1                   -74.002472 2013-09-12 18:48:34             236   
2                   -73.991255 2013-09-18 19:50:05             513   
3                   -73.974347 2013-09-28 12:03:58             441   
4                   -73.994811 2013-09-07 18:46:38             476   
...                        ...                 ...             ...   
49995               -73.991255 2013-09-21 20:08:59             478   
49996               -73.983625 2013-09-14 17:52:17             151   
49997               -73.989551 2013-09-07 18:25:51             524   
49998               -73.978058 2013-09-07 14:14:04             373   
49999               -73.994618 2013-09-23 22:00:46             488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year  \
0                 -73.997262            993      Subscriber      1968.0   
1                 -73.987140            581      Subscriber      1983.0   
2                 -73.988639            361      Subscriber      1989.0   
3                 -73.967416            561      Subscriber      1988.0   
4                 -73.979661           2296  Non-Subscriber      1976.0   
...                      ...            ...             ...         ...   
49995             -73.998842           2697  Non-Subscriber      1976.0   
49996             -73.997203           1191      Subscriber      1970.0   
49997             -73.983169           1050      Subscriber      1961.0   
49998             -73.953820           2697  Non-Subscriber      1976.0  

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 49977 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   bike_id                  49977 non-null  int64         
 1   weekday                  49977 non-null  object        
 2   start_hour               49977 non-null  int64         
 3   start_time               49977 non-null  datetime64[ns]
 4   start_station_id         49977 non-null  int64         
 5   start_station_name       49977 non-null  object        
 6   start_station_latitude   49977 non-null  float64       
 7   start_station_longitude  49977 non-null  float64       
 8   end_time                 49977 non-null  datetime64[ns]
 9   end_station_id           49977 non-null  int64         
 10  end_station_name         49977 non-null  object        
 11  end_station_latitude     49977 non-null  float64       
 12  end_station_longitude    49977 non-nu

In [38]:
# Changing start_station_id column from int to string
df['start_station_id'] = df['start_station_id'].astype('str')

In [39]:
# Changing end_station_id column from int to string
df['end_station_id'] = df['end_station_id'].astype('str')

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 49977 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   bike_id                  49977 non-null  int64         
 1   weekday                  49977 non-null  object        
 2   start_hour               49977 non-null  int64         
 3   start_time               49977 non-null  datetime64[ns]
 4   start_station_id         49977 non-null  object        
 5   start_station_name       49977 non-null  object        
 6   start_station_latitude   49977 non-null  float64       
 7   start_station_longitude  49977 non-null  float64       
 8   end_time                 49977 non-null  datetime64[ns]
 9   end_station_id           49977 non-null  object        
 10  end_station_name         49977 non-null  object        
 11  end_station_latitude     49977 non-null  float64       
 12  end_station_longitude    49977 non-nu

In [41]:
# Changing birth_year column from float to int
df['birth_year'] = df['birth_year'].astype('int64')

In [42]:
df

bike_id weekday  start_hour          start_time start_station_id  \
0        16013     Mon          18 2013-09-09 18:18:55              523   
1        15230     Thu          18 2013-09-12 18:38:53              257   
2        17942     Wed          19 2013-09-18 19:44:04              479   
3        19683     Sat          11 2013-09-28 11:54:37              527   
4        18024     Sat          18 2013-09-07 18:08:22              521   
...        ...     ...         ...                 ...              ...   
49995    17988     Sat          18 2013-09-21 18:39:20              479   
49996    18458     Sat          17 2013-09-14 17:32:26              261   
49997    15054     Sat          18 2013-09-07 18:08:21              402   
49998    16763     Sat          13 2013-09-07 13:25:35              271   
49999    16794     Mon          21 2013-09-23 21:54:36              515   

            start_station_name  start_station_latitude  \
0              W 38 St & 8 Ave               40.754666   
1      Lispenard St & Broadway               40.719392   
2              9 Ave & W 45 St               40.760193   
3              E 33 St & 1 Ave               40.743156   
4              8 Ave & W 31 St               40.750450   
...                        ...                     ...   
49995          9 Ave & W 45 St               40.760193   
49996     Johnson St & Gold St               40.694749   
49997       Broadway & E 22 St               40.740343   
49998   Ashland Pl & Hanson Pl               40.685282   
49999         W 43 St & 10 Ave               40.760094   

       start_station_longitude            end_time end_station_id  \
0                   -73.991382 2013-09-09 18:35:28            334   
1                   -74.002472 2013-09-12 18:48:34            236   
2                   -73.991255 2013-09-18 19:50:05            513   
3                   -73.974347 2013-09-28 12:03:58            441   
4                   -73.994811 2013-09-07 18:46:38            476   
...                        ...                 ...            ...   
49995               -73.991255 2013-09-21 20:08:59            478   
49996               -73.983625 2013-09-14 17:52:17            151   
49997               -73.989551 2013-09-07 18:25:51            524   
49998               -73.978058 2013-09-07 14:14:04            373   
49999               -73.994618 2013-09-23 22:00:46            488   

                   end_station_name  end_station_latitude  \
0                   W 20 St & 7 Ave             40.742388   
1               St Marks Pl & 2 Ave             40.728419   
2                  W 56 St & 10 Ave             40.768254   
3                   E 52 St & 2 Ave             40.756014   
4                   E 31 St & 3 Ave             40.743943   
...                             ...                   ...   
49995              11 Ave & W 41 St             40.760301   
49996      Cleveland Pl & Spring St             40.721816   
49997               W 43 St & 6 Ave             40.755273   
49998  Willoughby Ave & Walworth St             40.693317   
49999               W 39 St & 9 Ave             40.756458   

       end_station_longitude  trip_duration      subscriber  birth_year  \
0                 -73.997262            993      Subscriber        1968   
1                 -73.987140            581      Subscriber        1983   
2                 -73.988639            361      Subscriber        1989   
3                 -73.967416            561      Subscriber        1988   
4                 -73.979661           2296  Non-Subscriber        1976   
...                      ...            ...             ...         ...   
49995             -73.998842           2697  Non-Subscriber        1976   
49996             -73.997203           1191      Subscriber        1970   
49997             -73.983169           1050      Subscriber        1961   
49998             -73.953820           2697  Non-Subscriber        1976   
49999             -73.

# 04 Data consistency checks

In [43]:
df_dups = df[df.duplicated()]

In [44]:
df_dups

Empty DataFrame
Columns: [bike_id, weekday, start_hour, start_time, start_station_id, start_station_name, start_station_latitude, start_station_longitude, end_time, end_station_id, end_station_name, end_station_latitude, end_station_longitude, trip_duration, subscriber, birth_year, gender]
Index: []

## No duplicates found

In [45]:
# Checking for mixed data types
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

## No mixed data type columns found

In [47]:
df.head(30)

bike_id weekday  start_hour          start_time start_station_id  \
0     16013     Mon          18 2013-09-09 18:18:55              523   
1     15230     Thu          18 2013-09-12 18:38:53              257   
2     17942     Wed          19 2013-09-18 19:44:04              479   
3     19683     Sat          11 2013-09-28 11:54:37              527   
4     18024     Sat          18 2013-09-07 18:08:22              521   
5     14581     Fri          22 2013-09-27 22:01:01              293   
6     15022     Sun          13 2013-09-29 13:13:06              380   
7     16253     Sat          10 2013-09-21 10:48:14              346   
8     16745     Fri          12 2013-09-20 12:29:49              510   
9     19556     Tue          16 2013-09-03 16:56:06              446   
10    15732     Sat          20 2013-09-07 20:01:58              457   
11    15316     Fri          23 2013-09-06 23:02:53              251   
12    15522     Sun           2 2013-09-22 02:02:47              393   
13    15503     Fri           9 2013-09-13 09:09:15              388   
14    17172     Tue          15 2013-09-24 15:06:29              546   
15    15973     Mon          18 2013-09-16 18:45:05              167   
16    19355     Mon           6 2013-09-23 06:54:47              404   
17    15651     Mon          18 2013-09-30 18:32:48              363   
18    18863     Fri           7 2013-09-13 07:46:22              445   
19    18846     Tue          17 2013-09-24 17:48:27              505   
20    14814     Sun           1 2013-09-29 01:04:32              528   
21    18434     Sun           7 2013-09-08 07:42:00              512   
22    16918     Fri          20 2013-09-20 20:49:28              366   
23    18024     Thu          14 2013-09-05 14:59:43              380   
24    15136     Mon          18 2013-09-09 18:45:24              385   
25    17723     Fri           9 2013-09-13 09:59:21              252   
26    17506     Sun          16 2013-09-01 16:18:58              389   
27    17273     Fri           8 2013-09-13 08:39:07              526   
28    15578     Fri           8 2013-09-27 08:31:03              420   
29    20611     Wed          19 2013-09-11 19:33:51              503   

              start_station_name  start_station_latitude  \
0                W 38 St & 8 Ave               40.754666   
1        Lispenard St & Broadway               40.719392   
2                9 Ave & W 45 St               40.760193   
3                E 33 St & 1 Ave               40.743156   
4                8 Ave & W 31 St               40.750450   
5          Lafayette St & E 8 St               40.730287   
6               W 4 St & 7 Ave S               40.734011   
7            Bank St & Hudson St               40.736529   
8                W 51 St & 6 Ave               40.760660   
9                W 24 St & 7 Ave               40.744876   
10            Broadway & W 58 St               40.766953   
11           Mott St & Prince St               40.723180   
12             E 5 St & Avenue C               40.722992   
13              W 26 St & 10 Ave               40.749718   
14          E 30 St & Park Ave S               40.744449   
15               E 39 St & 3 Ave               40.748901   
16               9 Ave & W 14 St               40.740583   
17                West Thames St               40.708347   
18            E 10 St & Avenue A               40.727408   
19               6 Ave & W 33 St               40.749013   
20               2 Ave & E 31 St               40.742909   
21               W 29 St & 9 Ave               40.750073   
22      Clinton Ave & Myrtle Ave               40.693261   
23              W 4 St & 7 Ave S               40.734011   
24               E 55 St & 2 Ave               40.757973   
25  MacDougal St & Washington Sq               40.732264   
26           Broadway & Berry St               40.710446   
27               E 33 St & 5 Ave               40.747659   
28  Clermont Ave & Lafayette Av

In [48]:
# Exporting clean dataframe
df.to_csv(os.path.join(path, '02 Data', 'Prepared data', 'citibike.csv'))